### PR TITLE
Cuts the number of `apply_damage` copypaste procs from 3(.5) to 1, fixing a few bugs along the way

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

- `apply_damage` is now uniform, handled, in a single place. At the living level. 
- Fixes people being held at gunpoint from being unable to flinch from taking damage. 
- Fixes a few signals (explodable, glass jaw) of apply damage maybe potentially not having effects if the passed hit zone was a bodypart and not a def zone. 
- Fixes regenerator component always halting regeneration no matter what damage taken.
- Fixes pressure damage being unaffected by `brute_mod` despite what documentation claims. 
- Signalizes the `check_species_weakness` proc on species. 

## Why It's Good For The Game

The copy+paste of this proc made it awful, awful to maintain or add features to. And also made it very bug prone. 

This just takes a step forward in making sustaining damage a lot more consistent, not only across all mobs, but across all methods of... taking damage. Unarmed attacks, hulk, item attacks, etc. 

## Changelog

:cl: Melbert
fix: People held at gunpoint can now flinch when being hit. 
fix: Regenerating mobs no longer stop regenerating no matter hit with what. 
fix: Pressure damage is now properly modified by a mob's brute damage modifier. 
fix: Fixes some occasions which some effects (glass jaw, explodable worn items) won't respond to hits. 
refactor: Refactored core code related to mobs sustaining damage. 
/:cl:

### CODE REVIEW

**Comment by MrMelbert** [comment]
This one should probably be testmerged, maybe. Also thank god for the damage unit tests.

File: code/datums/components/regenerator.dm [hunk 1]
```diff
@@ -45,12 +45,10 @@
 		deltimer(regeneration_start_timer)
 
 /// When you take damage, reset the cooldown and start processing
-/datum/component/regenerator/proc/on_take_damage(datum/source, damage, damagetype)
+/datum/component/regenerator/proc/on_take_damage(datum/source, damage, damagetype, ...)
 	SIGNAL_HANDLER
 
-	if (damage <= 0)
-		return
-	if (locate(damagetype) in ignore_damage_types)
```
Comment (MrMelbert) [inline code/datums/components/regenerator.dm:53]: `locate("string") in list("string")` doesn't work.

File: code/modules/mob/living/carbon/human/species_types/zombies.dm [hunk 2]
```diff
@@ -149,20 +149,22 @@
...
 
 /datum/species/zombie/infectious/spec_stun(mob/living/carbon/human/H,amount)
 	. = min(20, amount)
 
-/datum/species/zombie/infectious/apply_damage(damage, damagetype = BRUTE, def_zone = null, blocked, mob/living/carbon/human/H, spread_damage = FALSE, forced = FALSE, wound_bonus = 0, bare_wound_bonus = 0, sharpness = NONE, attack_direction = null, attacking_item)
-	. = ..()
-	if(.)
+/datum/species/zombie/infectious/proc/damage_sustained(datum/source, damage_dealt, ...)
+	SIGNAL_HANDLER
+	if(damage_dealt > 0)
```
Comment (MrMelbert) [inline code/modules/mob/living/carbon/human/species_types/zombies.dm:167]: Infectious zombies should probably be using `/datum/component/regenerator`

**Comment by jlsnow301** [comment]
This is very likely going to be a balance change to bitrunning - but totally welcome

**Pull request merged**